### PR TITLE
adding login and request handlers

### LINKED ISSUE #6: As a developer, I want to implement a logging system

### LINKED ISSUE #10: As a developer, I want to implement tn account setup script and enable login on the backend

### PR DESCRIPTION
**Pull request by ghost**
fixes #10 
fixes #6

### CODE REVIEW

File: Backend/common.js [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+// <Global Errors> ------------------------------------------
+/**
+ * These errors are only for BE and API use
+ * Error codes and their corresponding message. Error codes are under different
+ * categories:
+ * 1000 -> system
+ * 2000 -> user
+ */
+
+const errors = Object.freeze({
```
Comment (ahmedqarmout) [inline Backend/common.js]: I forgot to add these to the UI, will fix this

File: Backend/users.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        if (err) {
+            return callback(common.getError(1002), null);
+        }
+
+        const currentDate = common.getDate();
+        var userToAdd = {};
+
+        userToAdd._id = common.getUUID();
+        userToAdd.username = user.username.toLowerCase();
+        userToAdd.fname = user.fname;
```
Comment (ahmedqarmout) [inline Backend/users.js]: need to add themes, and permissions

File: Backend/common.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+exports.variableTypes = variableTypes;
+// </Global Constants> ------------------------------------------
+
+// <Global Function> --------------------------------------------
+/**
+ * get a unique string
+ *
+ * @return {string}
+ */
+var getUUID = function () {
```
Comment (ahmedqarmout) [inline Backend/common.js]: var -> const of all functions

File: Backend/common.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+ * Error codes and their corresponding message. Error codes are under different
+ * categories:
+ * 1000 -> system
+ * 2000 -> user
+ */
+
+const errors = Object.freeze({
+    //1000 system
+    1000: 'invalid request',
+    1001: 'failed to connet to the db',
```
Comment (SergeyGV) [inline Backend/common.js]: connet -> connect

File: Backend/setup.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+    if (password !== password2) {
+        logger.error('Your passwords do not match, please try again.');
+        process.exit(1);
+    }
+
+    const user = {
+        username: username,
+        password: password,
+        fname: username,
```
Comment (SergeyGV) [inline Backend/setup.js]: Setting fname and lname to username is just temporary here, right?

**Review by SergeyGV — APPROVED** [review]
Nice stuff

File: Backend/setup.js (hunk 5 above)
Comment (ahmedqarmout) [reply Backend/setup.js]: yeah, they can edit that later from their profile page. Or we can fix that in here

Comment (SergeyGV) [reply Backend/setup.js]: Preferrable to have it fixed here, it's an odd thing to have to edit

Comment (ahmedqarmout) [reply Backend/setup.js]: fixed it. now it prompts for the first and last name

File: Backend/common.js (hunk 3 above)
Comment (ahmedqarmout) [reply Backend/common.js]: fixed

File: Backend/common.js (hunk 4 above)
Comment (ahmedqarmout) [reply Backend/common.js]: fixed

File: Backend/users.js (hunk 2 above)
Comment (ahmedqarmout) [reply Backend/users.js]: fixed

File: Backend/common.js (hunk 1 above)
Comment (ahmedqarmout) [reply Backend/common.js]: fixed

File: Backend/logger.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+/**
+ * initiate the logging with file
+ */
+const init = function () {
+    const currentDate = common.getDateFormatted('YYYY-MM-DD');
+    if (!process.env.DEBUG && dateStamp !== currentDate) {
+        dateStamp = currentDate;
+        const logger = fs.createWriteStream(`${__dirname}/../Logs/${dateStamp}.log`, { 'flags': 'a' });
+        process.stdout.write = process.stderr.write = logger.write.bind(logger);
+        process.on('uncaughtException', function (err) {
```
Comment (esteeroramy) [inline Backend/logger.js]: add an empty line for reading clarity

File: Backend/logger.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    init();
+    console.error(`[{${common.getDate()}}] ${type}: ${text}`);
+}
+
+/**
+ * initiate the logging with file
+ */
+const init = function () {
+    const currentDate = common.getDateFormatted('YYYY-MM-DD');
+    if (!process.env.DEBUG && dateStamp !== currentDate) {
```
Comment (esteeroramy) [inline Backend/logger.js]: add an empty line for reading clarity

File: Backend/setup.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+const rls = require('readline-sync');
```
Comment (esteeroramy) [inline Backend/setup.js]: can we sort these alphabetically

File: Backend/setup.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    const password = rls.question('Enter your password: ', {
+        hideEchoBack: true,
+        mask: '*'
+    });
+    const password2 = rls.question('Confirm your password: ', {
+        hideEchoBack: true,
+        mask: '*'
+    });
+
+    if (password !== password2) {
```
Comment (esteeroramy) [inline Backend/setup.js]: we should probably not end their setup if the passwords don't match, we should have this in a do while loop until the 2 passwords match

File: Templates/navbar.pug [hunk 10]
```diff
@@ -7,10 +7,10 @@
             ul.left.hide-on-med-and-down
                 li#nav-projects
                     a.navbarLink(href='#!') #{__('pageNames.projects')}
-                if canAccessUsers
+                if canAccessUsersList
```
Comment (esteeroramy) [inline Templates/navbar.pug]: I dont think this should be changed to `list`, it doesnt reflect the correct behaviour. also done you need `user.canAccess.....`

Comment (esteeroramy) [reply Templates/navbar.pug]: same with setings

File: velocity.js [hunk 11]
```diff
@@ -1,28 +1,52 @@
...
+const logger = require('./Backend/logger.js');
+const users = require('./Backend/users.js');
+
 const app = express();
 
 // File names to render
-const login = 'login';
+const loginPage = 'login';
 const modeSelectorPage = 'modeSelector';
+const pageNotFound = 'pageNotFound';
```
Comment (esteeroramy) [inline velocity.js]: `pageNotFound  -> `pageNotFoundPage`

**Review by esteeroramy — APPROVED** [review]

**Pull request merged**